src/scrub.S

```diff
@@ -0,0 +1,49 @@
+/* Copyright 2020 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/*
+ * Scrub memory with zero
+ */
+.type _metal_memory_scrub, @function
+_metal_memory_scrub:
+    /* Zero out memory if exist. */
+    bge t1, t2, 2f
+1:
+#if __riscv_xlen == 32
+    sw   x0, 0(t1)
+    addi t1, t1, 4
+    blt  t1, t2, 1b
+#else
+    sd   x0, 0(t1)
+    addi t1, t1, 8
+    blt  t1, t2, 1b
+#endif
+2:
+    ret
+
+/*
+ * Initialize all memories available to zero
+ * This must be call before seting up any stack(s)
+ */
+.global __metal_before_start
+.type __metal_before_start, @function
+__metal_before_start:
```

Comment:
you'll need to save and restore 'ra' across this function.

Reply:
This get called using val though. It is a single instruction to jump and save return address: jump and link (jal)